### PR TITLE
Posexional row as struct

### PR DESCRIPTION
**Pull request by GPrimola**
This is a proposal for #11.

**Obs 1**: There's **no breaking changes**.
**Obs 2**: I apologize about the size of this PR. 😅

## Major Changes
- Option `[:as_struct]` to `PosexionalRow.__using__/1` macro. **Not breaking**.
- New field `Posexional.Field.Field` that parses field type text. It behaves like `Ecto.Schema.field/3`. **Not breaking**.

### `PosexionalRow.__using__/1` with option `[:as_struct]` and `PosexionalRow.field/4` with simple data types:
Suppose we have the following row definition to be mapped to a struct:

```elixir
defmodule Row1 do
  use PosexionalRow, [:as_struct]

  guesser &__MODULE__.guesser/1

  fixed_value "row1"
  field :id, :id, 3
  field :binary_id, :binary_id, 3
  field :integer, :integer, 2
  field :float, :float, 6
  field :boolean, :boolean, 5
  field :string, :string, 4
  field :binary, :binary, 4

  def guesser("row1" <> _), do: true
  def guesser(_), do: false
end

defmodule StructRowFile do
  use PosexionalFile

  row Row1
end
```

and a positional file (example.txt - not included) with the content `row1123ABA00432.99falseJohn Doe`. Then

```elixir
StructRowFile.read("row1123ABA00432.99falseJohn Doe")
=>
[
  %Row1{
    binary: "Doe",
    binary_id: "ABA",
    boolean: false,
    float: 432.99,
    id: 123,
    integer: 0,
    string: "John"
  }
]
```

### `PosexionalRow.__using__/1` with option `[:as_struct]` and `PosexionalRow.field/4` with complex data types:

```elixir
defmodule Row2 do
  use PosexionalRow, [:as_struct]

  guesser &__MODULE__.guesser/1

  field :custom, :custom, 4, parser: &String.to_atom/1
  field :array, :array, 4, separator: ""
  field :list, :list, 7, separator: "|"
  field :decimal, :decimal, 8
  field :date, :date, 8, format_in: "{0M}{0D}{YYYY}"
  field :time, :time, 8, format_in: "%H%M%S%f"
  field :time2, :time, 6, format_in: "%H%M%S"
  field :datetime, :datetime, 14, format_in: "%d%m%Y%H%M%S", timezone: "America/Sao_Paulo"
  field :naive_dt, :naive_datetime, 23, format_in: "%d/%m/%Y@%H:%M:%S.%L"
  field :utc_dt, :utc_datetime, 23, format_in: "%Y-%m-%d %H:%M:%S.%L"

  def guesser("afoo" <> _), do: true
  def guesser(_), do: false
end

defmodule StructRowFile do
  use PosexionalFile

  row Row2
end

StructRowFile.read("afoo1234cat|dog123.45  01312021235959990745001209198807450031/12/2021@23:59:59.9992022-01-01 00:00:00.000")
=>
[
%Row2{
     array: ["1", "2", "3", "4"],
     custom: :afoo,
     date: ~D[2021-01-31],
     datetime: %DateTime{
       year: 1988,
       month: 9,
       day: 12,
       hour: 7,
       minute: 45,
       second: 00,
       utc_offset: -10800,
       time_zone: "America/Sao_Paulo",
       std_offset: 0,
       zone_abbr: "-03"
     },
     decimal: 123.45,
     list: ["cat", "dog"],
     naive_dt: ~N[2021-12-31 23:59:59.999],
     time: ~T[23:59:59.99],
     time2: ~T[07:45:00],
     utc_dt: ~U[2022-01-01 00:00:00Z]
   }
]
```

If you want more information about the behaviour, you can check the `test/posexional/field/field_test.exs` file where there are examples of bad input and how it's been handled.

Please, feel free to comment, propose changes, code refactor, variable, module, function and option naming, etc. I'll be glad to tackle everything!

Congrats about the lib, though! It's been very helpful! 😄

### CODE REVIEW

File: lib/posexional/field/field.ex [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    _ ->
+      raise ArgumentError, "#{value_str} could not be parsed as datetime with format #{format_in}"
+  end
+
+  defp parse(value_str, :utc_datetime, opts), do: parse(value_str, :datetime, Map.put(opts, :timezone, :utc))
+
+  defp parse(value_str, type, _opts),
+    do: raise(ArgumentError, "Provide a :parser option as a function to customize type #{type} to value #{value_str}")
+
+  @spec read(__MODULE__.t(), binary) :: any()
```
Comment (neslinesli93) [inline lib/posexional/field/field.ex:156]: You can omit `__MODULE__` here, `t()` is enough

File: lib/posexional/row.ex [hunk 2]
```diff
@@ -132,7 +133,9 @@ defmodule Posexional.Row do
       |> elem(0)
       |> Enum.filter(fn {k, _} -> not (k in [:empty_field]) end)
 
-    [{name, res}]
+    if not is_nil(struct_module),
+      do: [struct(struct_module, res)],
```
Comment (neslinesli93) [inline lib/posexional/row.ex:137]: Maybe we should keep the same interface to the outside world? So a tuple with `{name, struct}`

**Review by neslinesli93 — CHANGES_REQUESTED** [review]
Amazing work, for real :heart: 

Unfortunately, our CI service is purposely disabled for public projects, so it can't be run on this PR :cry: 
I've tried locally to run credo and dialyzer and I got a bunch of errors. You can try `mix check` and see that credo complains about single pipes, and some other stuff.

Apart from that, I'm not a huge fan of having a `Posexional.Field.Field` module, mostly for the repetition of words and because imho it's not that self-explanatory. Maybe something like `TypedField` would be better? I like the reference to Ecto, but the repetition kills me :D

**Comment by GPrimola** [comment]
Hey @neslinesli93,

Thank you for your review! I've already tackled everything you pointed out!
And agree, `Posexional.Field.Field` was a bad name! Thanks for the suggestion! ;)

If you have anything more to add, please feel free to do it! :)

File: test/posexional/field/field_test.exs [hunk 3]
```diff
@@ -0,0 +1,214 @@
+defmodule Posexional.Field.FieldTest do
+  use Posexional.Case, async: true
+
+  defmodule Row1 do
+    use PosexionalRow, [:as_struct]
+
+    guesser &__MODULE__.guesser/1
+
+    fixed_value("row1")
```
Comment (chess4ever) [inline test/posexional/field/field_test.exs]: does it make sense to avoid the use of parentheses as for ecto schema, since it is a macro?

**Review by chess4ever — APPROVED** [review]

File: test/posexional/field/field_test.exs (hunk 3 above)
Comment (chess4ever) [reply test/posexional/field/field_test.exs]: sorry it is outdated, I will resolve it

**Comment by chess4ever** [comment]
Hi @GPrimola ! You have done a really great work here! Thank you for you contribution!
We would like to merge your PR a soon as possible since it is valuable.

I have run dialyzer locally (in the container) and there are still some issues (I count 9 of those).

Let us know if you need help to solve them! We are glad to help!

Thank you again!

**Review by chess4ever — APPROVED** [review]

**Comment by chess4ever** [comment]
@GPrimola is it ready to be merged?

**Comment by GPrimola** [comment]
Hey @chess4ever, thanks!

Actually I missed letting dialyzer run fully.
Now I think everything should be fine!

Thank you!

**Pull request merged**